core/modules/qos.h

```diff
@@ -0,0 +1,125 @@
+// Copyright Intel Corp.
+// All rights reserved
+//
+// Redistribution and use in source and binary forms, with or without
+// modification, are permitted provided that the following conditions are met:
+//
+// * Redistributions of source code must retain the above copyright notice, this
+// list of conditions and the following disclaimer.
+//
+// * Redistributions in binary form must reproduce the above copyright notice,
+// this list of conditions and the following disclaimer in the documentation
+// and/or other materials provided with the distribution.
+//
+// * Neither the names of the copyright holders nor the names of their
+// contributors may be used to endorse or promote products derived from this
+// software without specific prior written permission.
+//
+// THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS IS"
+// AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
+// IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE
+// ARE DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT HOLDER OR CONTRIBUTORS BE
+// LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR
+// CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF
+// SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS
+// INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN
+// CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE)
+// ARISING IN ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE
+// POSSIBILITY OF SUCH DAMAGE.
+
+#ifndef BESS_MODULES_QOS_H_
+#define BESS_MODULES_QOS_H_
+
+#include "../module.h"
+
+#include <rte_config.h>
+#include <rte_hash_crc.h>
+
+#include "../pb/module_msg.pb.h"
+#include "../utils/metering.h"
+
+using bess::utils::Metering;
+using bess::utils::MeteringKey;
+
+#define MAX_FIELDS 8
+#define MAX_FIELD_SIZE 8
+static_assert(MAX_FIELD_SIZE <= sizeof(uint64_t),
+              "field cannot be larger than 8 bytes");
+
+#define HASH_KEY_SIZE (MAX_FIELDS * MAX_FIELD_SIZE)
+
+#if __BYTE_ORDER__ != __ORDER_LITTLE_ENDIAN__
+#error this code assumes little endian architecture (x86)
+#endif
+
+enum { FieldType = 0, ValueType };
+/*
+struct QosKey {
+  uint64_t u64_arr[MAX_FIELDS];
+};
+*/
+struct QosData {
+  uint8_t qfi;
+  uint32_t cir;
+  uint32_t pir;
+  uint32_t cbs;
+  uint32_t pbs;
+  uint32_t ebs;
```

Comment:
NOTE: DPDK structs have these as uint64_t. We might have to change at other places too. On the other hand just with uint32_t we can go upto 34,359,738,360 bps i.e., 34 Gbps (a lot)? Check this math.